external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1812 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(not(feature = "gas-profiler"))]
+pub(crate) use vm_tracer_noop::*;
+mod vm_tracer_noop {
+    use move_trace_format::format::MoveTraceBuilder;
+
+    #[allow(dead_code)]
+    pub(crate) struct VMTracer<'a> {
+        _phantom: std::marker::PhantomData<&'a ()>,
+    }
+
+    impl<'a> VMTracer<'a> {
+        #[allow(dead_code)]
+        pub(crate) fn new(_trace: &'a mut MoveTraceBuilder) -> Self {
+            Self {
+                _phantom: std::marker::PhantomData,
+            }
+        }
+    }
+}
+
+#[cfg(feature = "gas-profiler")]
+pub(crate) use vm_tracer_impl::*;
+#[cfg(feature = "gas-profiler")]
+mod vm_tracer_impl {
+    use crate::{
+        interpreter::{Frame, Interpreter},
+        loader::{Function, Loader},
+    };
+    use move_binary_format::{
+        errors::PartialVMError,
+        file_format::{ConstantPoolIndex, SignatureIndex},
+    };
+    use move_core_types::{
+        account_address::AccountAddress,
+        annotated_value::{MoveTypeLayout, MoveValue},
+        language_storage::TypeTag,
+    };
+    use move_trace_format::format::{
+        DataLoad, Effect as EF, Location, MoveTraceBuilder, Read, RefType, TraceIndex, TraceValue,
+        TypeTagWithRefs, Write,
+    };
+    use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+    use std::collections::BTreeMap;
+
+    /// Internal state for the tracer. This is where the actual tracing logic is implemented.
+    pub(crate) struct VMTracer<'a> {
+        trace: &'a mut MoveTraceBuilder,
+        link_context: Option<AccountAddress>,
+        pc: Option<u16>,
+        active_frames: BTreeMap<TraceIndex, FrameInfo>,
```

Comment:
Note that this is sparse (this is keyed by `TraceIndex` and not e.g., call stack depth or anything like that), so moving to a `Vec<...>` is probably not a good idea here...

Reply:
maybe I did not understand this too well :) and here is what I thought:
`BTreeMap<TraceIndex, FrameInfo>` is acting as a stack whenever we get  the latest value (top of the stack, `last_key_value`) and it works because `TraceIndex` is guaranteed to be monotonically growing and as such a "later" frame has always a higher value of a frame up the stack (previous frame).
However we also use `TraceIndex` as a key into the "stack" to get roots and such.
Is that the case?
If that was the case (if what I wrote is true), a `Vec<(TraceIndex, FrameInfo)>` or `Vec<FrameInfo>` if `FrameInfo` had the `TraceIndex` would work as well with a binary search in place of all the `get` in the code. All the usage of a stack (`last_key_value`) would simply work off a vector which is more natural.
In other words (again assuming I understood anything) the perf characteristics of a vector with a binary search would be the same of the `BTreeMap` for the `get`(log(n)) and better otherwise in terms of space and time.

All in all at the end is not a big deal in terms of perf, as call stacks are not really that deep, and maybe better when reading the code?

Did I got the usage right?

Reply:
You got the usage right!

I think it's better to keep it as a `BTreeMap` as that feels more "natural" to me at least when reading/reasoning about the code since you generally need random access into the stack more frequently then most recent call frame (mainly due to mutating references). Agreed that it's not that big a deal either way though (and we can convert at a later date if we want/need to). 

If we wanted to we could also move to an [`IndexMap`](https://docs.rs/indexmap/latest/indexmap/) for this as well which would give the ideal access patterns for this -- `O(1)` random access and `O(1)` "last frame"/call-stack order operations.

Reply:
fine keeping things as they are and I still think a vector would be the ideal struct - lol